cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/AbstractNotificationSubscriberService.java

```diff
@@ -228,29 +201,71 @@ private String fetchAndProcessNotifications(DatasetContext context, MessageFetch
             lastFetchedMessageId = message.getId(); // update messageId to skip this message in next fetch
             continue;
           }
-          updateJobQueue(context, notification);
+          processNotification(context, notification);
           lastFetchedMessageId = message.getId();
         }
 
+        // Persist the last fetched message id
         if (lastFetchedMessageId != null) {
-          jobQueue.persistSubscriberState(topic, lastFetchedMessageId);
+          updateMessageId(lastFetchedMessageId);
         }
 
         return lastFetchedMessageId;
       }
     }
 
-    abstract void updateJobQueue(DatasetContext context, Notification notification) throws Exception;
+    /**
+     * Loads the message id from storage. Note that this method is already executed inside a transaction.
+     *
+     * @return the message id, or null if no message id has been stored
+     */
+    public abstract String loadMessageId();
+
+    /**
+     * Persists the message id to storage. Note that this method is already executed inside a transaction.
+     */
+    public abstract void updateMessageId(String lastFetchedMessageId);
+
+    /**
+     * Processes the fetched notification
+     *
+     * @param context the dataset context
+     * @param notification the decoded notification
+     * @throws Exception
+     */
+    public abstract void processNotification(DatasetContext context, Notification notification) throws Exception;
   }
 
-  private class SchedulerEventNotificationSubscriberThread extends NotificationSubscriberThread {
+  /**
+   * Thread that subscribes to TMS notifications and adds the notification containing the schedule id to the job queue
+   */
+  protected class SchedulerEventNotificationSubscriberThread extends NotificationSubscriberThread {
```

Comment:
Yes, but the job queue uses the `multiThreadDatasetCache`, which only exists in the `AbstractNotificationSubscriberService`. I didn't want that to be exposed to the `ScheduleNotificationSubscriberService` since the `multiThreadDatasetCache` also used to create the transactional object.